omod/src/test/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProviderWebTest.java

```diff
@@ -71,43 +85,83 @@ public void getDiagnosticReportByWrongUuid_shouldReturn404() throws Exception {
+		when(service.updateDiagnosticReport(anyString(), any(DiagnosticReport.class))).thenReturn(diagnosticReport);
+		
+		MockHttpServletResponse response = put("/DiagnosticReport/" + DIAGNOSTIC_REPORT_UUID)
+		        .jsonContent(jsonDiagnosticReport).accept(FhirMediaTypes.JSON).go();
 		
 		assertThat(response, isOk());
 	}
 	
 	@Test
 	public void updateDiagnosticReport_shouldErrorForIdMismatch() throws Exception {
-		DiagnosticReport diagnosticReport = new DiagnosticReport();
-		diagnosticReport.setId(DIAGNOSTIC_REPORT_UUID);
+		String jsonDiagnosticReport;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_DIAGNOSTIC_REPORT_PATH)) {
+			jsonDiagnosticReport = IOUtils.toString(is);
+		}
 		
-		MockHttpServletResponse response = put("/DiagnosticReport/" + WRONG_UUID).accept(FhirMediaTypes.JSON).go();
+		MockHttpServletResponse response = put("/DiagnosticReport/" + WRONG_UUID).jsonContent(jsonDiagnosticReport)
+		        .accept(FhirMediaTypes.JSON).go();
 		
-		// TODO: Fix so status is 400, not 200. See
-		assertThat(response, isOk());
+		assertThat(response, isBadRequest());
+		assertThat(response.getContentAsString(),
+		    containsStringIgnoringCase("body must contain an ID element which matches the request URL"));
 	}
 	
 	@Test
 	public void updateDiagnosticReport_shouldErrorForNoId() throws Exception {
-		DiagnosticReport diagnosticReport = new DiagnosticReport();
-		diagnosticReport.setId(DIAGNOSTIC_REPORT_UUID);
+		String jsonDiagnosticReport;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_DIAGNOSTIC_REPORT_NoId_PATH)) {
+			jsonDiagnosticReport = IOUtils.toString(is);
+		}
 		
-		MockHttpServletResponse response = put("/DiagnosticReport/").accept(FhirMediaTypes.JSON).go();
+		MockHttpServletResponse response = put("/DiagnosticReport/" + DIAGNOSTIC_REPORT_UUID)
+		        .jsonContent(jsonDiagnosticReport).accept(FhirMediaTypes.JSON).go();
 		
-		// TODO: Fix so status is 400, not 200.
-		assertThat(response, isOk());
+		assertThat(response, isBadRequest());
+		assertThat(response.getContentAsString(), containsStringIgnoringCase("body must contain an ID element for update"));
+	}
+	
+	@Test
+	public void updateDiagnosticReport_shouldErrorForNonexistentDiagnosticReport() throws Exception {
+		String jsonDiagnosticReport;
+		try (InputStream is = this.getClass().getClassLoader().getResourceAsStream(JSON_DIAGNOSTIC_REPORT_WRONG_UUID_PATH)) {
+			jsonDiagnosticReport = IOUtils.toString(is);
+		}
+		
+		when(service.updateDiagnosticReport(eq(WRONG_UUID), any(DiagnosticReport.class)))
+		        .thenThrow(MethodNotAllowedException.class);
+		
+		MockHttpServletResponse response = put("/DiagnosticReport/" + WRONG_UUID).jsonContent(jsonDiagnosticReport)
+		        .accept(FhirMediaTypes.JSON).go();
+		
+		// Not sure why this is returning HTTP status code 0
+		assertThat(response, isMethodNotAllowed());
```

Comment:
I can't figure out why this test (and the corresponding Task Web test) return status code 0 instead of 405. The `MethodNotAllowedException` seems to not be getting handled correctly, and I couldn't untangle it after looking into it a bit.

Reply:
I'd need to dig into this a bit. 0 is not an HTTP status code...

Reply:
Ok. The problem here is a bit subtle, but because `MethodNotAllowedException` has no default constructor (i.e., a constructor that takes no arguments), there's a deeper error happening causes the wrong exception to be thrown causing the weird return code. To fix this, simple change the `.thenThrow()` part to something like `.thenThrow(new MethodNotAllowedException("DiagnosticReport " + WRONG_UUID + " does not exist"))` and the test will work as expected.